File: FirstAndLastElement.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
// Binary Search to find the first index and binary search to find the last index with the help of first index

public class FirstAndLastElement {
    private int binarySearch(int[] nums, int low, int high, int target,
                             boolean isFirstIndex) {
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (isFirstIndex) {
                if (nums[mid] == target) {
                    if (mid == low || nums[mid] != nums[mid - 1]) {
                        return mid;
                    } else {
                        high = mid - 1;
                    }
                } else if (nums[mid] > target) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else {
                if (nums[mid] == target) {
                    if (mid == high || nums[mid] != nums[mid + 1]) {
                        return mid;
                    } else {
                        low = mid + 1;
                    }
                } else if (nums[mid] < target) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
        }
        return -1;
    }

    public int[] searchRange(int[] nums, int target) {
        if (nums == null) return new int[]{-1, -1};

        int length = nums.length;
        if (length == 0) return new int[]{-1, -1};
        if (nums[0] > target && nums[length - 1] < target) {
            return new int[]{-1, -1};
        }

        // to find first index, BS for entire array
        int firstIndex = binarySearch(nums, 0, length - 1, target, true);
        if (firstIndex == -1) return new int[]{-1, -1};
        // to find second index, BS from first index to last element in the array
        int secondIndex = binarySearch(nums, firstIndex, length - 1, target, false);
        return new int[]{firstIndex, secondIndex};
    }
}

File: first_and_last_element.py
#time_complexit = lon(n)
#space_complexity = o(1)
#first here I am finding the lower index of target and for that I am using function and then for higher index another function. By using thsese I reduced the time complexity and also learned approach of using function instead of loops to reduce complexity.
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        if nums == None or len(nums) ==0:
            return [-1,-1]
        first = self.binarysearchfirst(nums,target)
        if first == -1:
            return [-1,-1]
        last = self.binarysearchlast(nums,target)
        return [first,last]
    
    def binarysearchfirst(self, nums: List[int], target: int) -> List[int]:
        low = 0
        high =len(nums)-1
        while low<=high:
            mid = low + (high - low)//2
            if nums[mid] == target:
                if (mid == 0) or (nums[mid] != nums[mid-1]):
                    return mid
                else:
                    high = mid-1
            elif nums[mid]<target:
                low = mid + 1
            else:
                high = mid -1
        return -1
    def binarysearchlast(self, nums: List[int], target: int) -> List[int]:
        low = 0
        high =len(nums)-1
        while low<=high:
            mid = low + (high - low)//2
            if nums[mid] == target:
                if (mid == len(nums)-1) or (nums[mid] != nums[mid+1]):
                    return mid
                else:
                    low = mid+1
            elif nums[mid]<target:
                low = mid + 1
            else:
                high = mid -1
        return -1

        